File: lib/beakvision.ts
export type BeakVisionMode = "mobile" | "computer" | "ground";

export type Point = { x: number; y: number };

export type ParsedAction = {
  type: string;
  point: Point;
  end_point?: Point;
  text?: string;
  direction?: string;
  key?: string;
  thought?: string;
};

export type ParseResponse = {
  success: boolean;
  data: {
    mode: BeakVisionMode;
    screen_description: string;
    layout_summary: string;
    elements: Array<{
      id: number;
      type: string;
      label: string;
      description: string;
      bounds: { x: number; y: number; width: number; height: number };
      center: Point;
      interactable: boolean;
      action_hint: string;
    }>;
    suggested_actions: string[];
    action?: ParsedAction;
    coordinate_space: { width: number; height: number };
  };
  meta?: Record<string, unknown>;
};

function required(name: string): string {
  const value = process.env[name];
  if (!value) {
    throw new Error(`Missing required environment variable: ${name}`);
  }
  return value;
}

export function parseUrl(): string {
  return required("BEAKVISION_PARSE_URL");
}

export function apiKey(): string {
  return required("BEAKVISION_API_KEY");
}

export async function parseScreenshot(params: {
  imageBase64: string;
  mode: BeakVisionMode;
  goal: string;
  context?: string;
}): Promise<ParseResponse> {
  const response = await fetch(parseUrl(), {
    method: "POST",
    headers: {
      "content-type": "application/json",
      authorization: `Bearer ${apiKey()}`,
    },
    body: JSON.stringify({
      image: params.imageBase64,
      mode: params.mode,
      goal: params.goal,
      context: params.context,
    }),
  });

  if (!response.ok) {
    const body = await response.text();
    throw new Error(`BeakVision request failed (${response.status}): ${body}`);
  }

  return (await response.json()) as ParseResponse;
}

File: lib/beakvision.py
from __future__ import annotations

import os
from typing import Any, Dict, Optional

import requests


def _required_env(name: str) -> str:
    value = os.getenv(name)
    if not value:
        raise RuntimeError(f"Missing required environment variable: {name}")
    return value


def parse_screenshot(
    *,
    image_base64: str,
    mode: str,
    goal: str,
    context: Optional[str] = None,
) -> Dict[str, Any]:
    response = requests.post(
        _required_env("BEAKVISION_PARSE_URL"),
        headers={
            "Authorization": f"Bearer {_required_env('BEAKVISION_API_KEY')}",
            "Content-Type": "application/json",
        },
        json={
            "image": image_base64,
            "mode": mode,
            "goal": goal,
            "context": context,
        },
        timeout=60,
    )
    response.raise_for_status()
    return response.json()
